File: leetcode/46-permute.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        if(nums == null || nums.length == 0) return res;
        backtrack(res, new ArrayList<Integer>(), nums);
        return res;
    }
    private void backtrack(List<List<Integer>> list, List<Integer> item, int[] nums) {
        if(item.size() == nums.length)
            list.add(new ArrayList<>(item));
        else {
            for(int i = 0; i < nums.length; i++) {
                if(item.contains(nums[i])) continue;
                item.add(nums[i]);
                backtrack(list, item, nums);
                item.remove(item.size()-1);
            }
        }
    }
}

File: leetcode/46-permute.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        permuteRecursive(nums, 0, res);
        return res;
    }
    void permuteRecursive(vector<int>& num, int begin, vector<vector<int>>& res) {
        if(begin >= num.size()) {
            res.push_back(num);
            return;
        }
        for(int i = begin; i < num.size(); i++) {
            swap(num[begin], num[i]);
            permuteRecursive(num, begin+1, res);
            swap(num[begin], num[i]);
        }
    }
};
